### ISSUE TITLE
RdKafka\ConsumerTopic->consumeStop takes forever

### ISSUE THREAD

**Issue by remizyaka** [php-rdkafka/php-rdkafka#115]
* PHP version: 5.6.30
 * librdkafka version: 0.9.5
 * php-rdkfaka version: 3.0.1

I'm trying to stop consumer  to save CPU resources when there are no new messages in queue and I'm sending process to sleep. But calling consumeStop method takes forever and process just hangs on it.

**Comment by ncreea** [comment]
Same issue, same versions (Centos 7 remi repo)
PHP 5.6.30
librdkafka 0.9.5
php-rdkfaka 3.0.1

**Comment by AndrzejFR** [comment]
Experiencing same issue.
```
<?php

$conf = new RdKafka\Conf();
$conf->set('debug', 'generic,broker,queue,msg,protocol');

$rk = new RdKafka\Consumer($conf);
$rk->setLogLevel(LOG_DEBUG);
$rk->addBrokers('127.0.0.1');

$topic = $rk->newTopic("stop-test");

$partition = 0;
$topic->consumeStart($partition, RD_KAFKA_OFFSET_BEGINNING);

$count = 0;
while (true) {
    $msg = $topic->consume($partition, 1000);
    if ($msg && $msg->err) {
        break;
    } elseif ($msg) {
        $count++;
    }
}

$topic->consumeStop($partition);

echo 'Finished. Consumed '. $count;
```
It will get stuck at the end, but if we comment out `$topic->consumeStop($partition);` the script will terminate. Can we skip `consumeStop`? It's not called in the consumer examples.

```
PHP 5.6.30-10+deb.sury.org~xenial+2 (cli)
rdkafka

rdkafka support => enabled
version => 3.0.1
build date => May 18 2017 18:08:41
librdkafka version (runtime) => 0.9.5
librdkafka version (build) => 0.9.5.0
```

Btw it works on
```
PHP 5.6.22-1+donate.sury.org~trusty+1 (cli)
rdkafka

rdkafka support => enabled
version => 0.9.1
build date => May 30 2016 13:54:06
librdkafka version => 0.8.6.0
```

**Comment by arnaud-lb** [comment]
Thanks for your feedback. @ncreea @remizyaka could also test with librdkafka 0.8.x ?

**Comment by AndrzejFR** [comment]
I just tested it a little more for different versions.
It works ok on 
PHP 7.1.5 (cli) (built: May 13 2017 00:03:34) ( NTS )

Seems that only versions 3.0.* on php5.6 are affected. Version of `librdkafka` doesn't matter.

Here is docker file I have used to run tests, just update version numbers as needed in lines 1, 6, 8.

```
FROM php:5.6.30-cli
RUN apt-get update
RUN apt-get install -y git-core libssl-dev libsasl2-dev make python
RUN git clone https://github.com/edenhill/librdkafka.git
WORKDIR /librdkafka
RUN git checkout tags/v0.9.5
RUN ./configure && make && make install
RUN pecl install rdkafka-3.0.1 && docker-php-ext-enable rdkafka
ADD ./kafka.php /run/kafka.php
CMD ["php", "/run/kafka.php"]
```

`kafka.php` as I posted above, here for completnes

```
<?php

$conf = new RdKafka\Conf();
$conf->set('debug', 'generic,broker,queue,msg,protocol');

$rk = new RdKafka\Consumer($conf);
$rk->setLogLevel(LOG_DEBUG);
$rk->addBrokers('127.0.0.1');

$topic = $rk->newTopic("stop-test");

$partition = 0;
$topic->consumeStart($partition, RD_KAFKA_OFFSET_BEGINNING);

$count = 0;
while (true) {
    $msg = $topic->consume($partition, 1000);
    if ($msg && $msg->err) {
        break;
    } elseif ($msg) {
        $count++;
    }
}

$topic->consumeStop($partition);

echo 'Finished. Consumed '. $count;
```

You just need to set brokers to your own kafka cluster.

**Comment by AndrzejFR** [comment]
Just repeating question, how bad will it be not to call `consumeStop` in the meantime? What this method does?

**Comment by arnaud-lb** [comment]
Thank you for the detailed report, this is fixed in master.

consumeStop just tells the client to stop pulling for new messages in the partition. This should have virtually no impact on CPU usage.

**Comment by AndrzejFR** [comment]
Thanks for quick fix, I see you tagged it as 3.0.2, when can we expect it to be available from pecl?

**Comment by arnaud-lb** [comment]
Yes: https://pecl.php.net/package/rdkafka/3.0.2

**Comment by ncreea** [comment]
All right! Thank you!

**Comment by yourNameIsGood** [comment]
Seems like my problem.

I have this rpc calling problem after topic->consumeStart( );
And I assumed it was something about memory leaking due to consumer not stopped,
so I went ahead and try to add this topic->consumeStop($partition);

It just freeze there like 4ever ...

**Comment by arnaud-lb** [comment]
@yourNameIsGood: This issue is fixed in php-rdkafka 3.0.2
